### ISSUE TITLE
Use the ``classical_function`` compiler instead of Oracle components

### ISSUE THREAD

**Issue by Cryoris** [Qiskit/qiskit-terra#5714]
### What is the expected enhancement?

In the last release of Terra, the classical function compiler has been added which uses the `tweedledum` library to compile a classical function to a circuit. This compiler yields much more efficient circuits than the Oracle components `LogicalExpressionOracle` and `TruthTableOracle`. 

Oracle for the logical expression `not a or (b and c)`:
```python
@classical_function
def leo(a: Int1, b: Int1, c: Int1) -> Int1:
      return (not a) or (b and c)

circuit = leo.synth()
```
yields
```
q_0: ──■────o───────
       │    │
q_1: ──o────┼───────
       │    │
q_2: ──■────■───────
     ┌─┴─┐┌─┴─┐┌───┐
q_3: ┤ X ├┤ X ├┤ X ├
     └───┘└───┘└───┘
Optimized circuit count (basis set 'u x cx'): OrderedDict([('u', 29), ('cx', 26), ('x', 5)])
```
whereas 
```python
leo = LogicalExpressionOracle('~a | (b ^ c)', optimization=True)
circuit = leo.construct_circuit()
```
yields
```python
     ┌───┐     ┌───┐┌───┐                          ┌───┐
v_0: ┤ X ├──■──┤ X ├┤ X ├───────────────────────■──┤ X ├───────────────
     └───┘  │  └───┘├───┤     ┌───┐             │  └───┘┌───┐     ┌───┐
v_1: ───────┼────■──┤ X ├──■──┤ X ├─────────────┼────■──┤ X ├──■──┤ X ├
     ┌───┐  │    │  ├───┤  │  ├───┤             │    │  ├───┤  │  └───┘
v_2: ┤ X ├──┼────■──┤ X ├──■──┤ X ├─────────────┼────■──┤ X ├──■───────
     └───┘┌─┴─┐  │  ├───┤  │  └───┘      ┌───┐┌─┴─┐  │  └───┘  │
c_0: ─────┤ X ├──┼──┤ X ├──┼─────────■───┤ X ├┤ X ├──┼─────────┼───────
          └───┘┌─┴─┐├───┤  │         │   ├───┤└───┘┌─┴─┐       │
c_1: ──────────┤ X ├┤ X ├──┼─────────■───┤ X ├─────┤ X ├───────┼───────
               └───┘└───┘┌─┴─┐┌───┐  │   ├───┤     └───┘     ┌─┴─┐
c_2: ────────────────────┤ X ├┤ X ├──■───┤ X ├───────────────┤ X ├─────
     ┌───┐               └───┘└───┘┌─┴──┐└───┘               └───┘
o_0: ┤ X ├─────────────────────────┤0   ├──────────────────────────────
     └───┘                         │  X │
a_0: ──────────────────────────────┤1   ├──────────────────────────────
                                   └────┘
Optimized circuit count (basis set 'u x cx'): OrderedDict([('u', 48), ('cx', 38), ('x', 17)])
```

The truth table oracle for `a xor b`:
```python
@classical_function
def tto(a: Int1, b: Int1) -> Int1:
    return (not a and b) or (a and not b)

circuit = tto.synth()
```
yields
```python
q_0: ──■───────
       │
q_1: ──┼────■──
     ┌─┴─┐┌─┴─┐
q_2: ┤ X ├┤ X ├
     └───┘└───┘
Optimized circuit count (basis set 'u x cx'): OrderedDict([('cx', 2)])
```
whereas
```python
tto =TruthTableOracle('0110')
circuit = tto.construct_circuit()
```
yields
```python
          ┌───┐
v_0: ──■──┤ X ├──■──
     ┌─┴─┐└───┘  │
o_0: ┤ X ├───────┼──
     └───┘     ┌─┴─┐
o_1: ──────────┤ X ├
               └───┘
Optimized circuit count (basis set 'u x cx'): OrderedDict([('cx', 2), ('x', 1)])
```

Since the Grover algorithm now works on circuits as oracles, I think we should remove these oracle classes. Alternatively, if there's some convenient functionality we want to keep, we should either 
(1) augment the classical compiler to allow that functionality, or if not possible
(2) replace the oracle components by ``QuantumCircuit``s that leverage the classical function compiler internally.

**Comment by woodsp-ibm** [comment]
Since this is highly related to use in the algorithms I labelled it that way though the resultant objects are likely library entities

**Comment by Cryoris** [comment]
Closed via #5830.
